src/Storages/ProjectionsDescription.h

```diff
@@ -94,7 +97,7 @@ struct ProjectionDescription
 
     bool isPrimaryKeyColumnPossiblyWrappedInFunctions(const ASTPtr & node) const;
 
-    Block calculate(const Block & block, ContextPtr context) const;
+    Block calculate(const Block & block, ContextPtr context, const IColumnPermutation * perm_ptr = nullptr) const;
```

Comment:
I don't fully understand what is `perm_ptr` here and where it is used.
Can you add a comment, please?

Reply:
Sure. It's to correctly generate `_part_offset` column in projection during insertion.